File: app-front/consommables/src/services/ArticleService.js
import api from './api'

const ArticleService = {
  // Generic Article operations
  getAllArticles() {
    return api.get('/articles')
  },

  getArticleById(id) {
    return api.get(`/articles/${id}`)
  },

  getArticleByCodeBarre(codeBarre) {
    return api.get(`/articles/code/${codeBarre}`)
  },

  getArticlesByType(type) {
    return api.get(`/articles/type/${type}`)
  },

  getArticlesByFamille(famille) {
    return api.get(`/articles/famille/${famille}`)
  },

  getArticlesByFournisseur(fournisseur) {
    return api.get(`/articles/fournisseur/${fournisseur}`)
  },

  getAllTypes() {
    return api.get('/articles/types/all')
  },

  getAllFamilles() {
    return api.get('/articles/familles/all')
  },

  createArticle(article) {
    return api.post('/articles', article)
  },

  updateArticle(id, article) {
    return api.put(`/articles/${id}`, article)
  },

  deleteArticle(id) {
    return api.delete(`/articles/${id}`)
  },

  getArticleStats() {
    return api.get('/articles/stats')
  },

  // Consommables
  getAllConsommables() {
    return api.get('/articles/consommables/all')
  },

  // Cartouches
  getAllCartouches() {
    return api.get('/articles/cartouches/all')
  },

  getCartouchesByColor(couleur) {
    return api.get(`/articles/cartouches/couleur/${couleur}`)
  },

  // Meubles
  getAllMeubles() {
    return api.get('/articles/meubles/all')
  },

  getMeublesByMaterial(materiau) {
    return api.get(`/articles/meubles/materiau/${materiau}`)
  },

  // Bureaux
  getAllBureaux() {
    return api.get('/articles/bureaux/all')
  },

  // Chaises
  getAllChaises() {
    return api.get('/articles/chaises/all')
  },

  // Etageres
  getAllEtageres() {
    return api.get('/articles/etageres/all')
  },

  // Imprimantes
  getAllImprimantes() {
    return api.get('/articles/imprimantes/all')
  },

  getImprimantesByMarque(marque) {
    return api.get(`/articles/imprimantes/marque/${marque}`)
  },

  // Ordinateurs
  getAllOrdinateurs() {
    return api.get('/articles/ordinateurs/all')
  },

  getOrdinateursByMarque(marque) {
    return api.get(`/articles/ordinateurs/marque/${marque}`)
  },

  getOrdinateursByType(typeOrdinateur) {
    return api.get(`/articles/ordinateurs/type/${typeOrdinateur}`)
  },

  // Moniteurs
  getAllMoniteurs() {
    return api.get('/articles/moniteurs/all')
  },

  getMoniteursByMarque(marque) {
    return api.get(`/articles/moniteurs/marque/${marque}`)
  },

  // Peripheriques
  getAllPeripheriques() {
    return api.get('/articles/peripheriques/all')
  },

  getPeripheriquesByType(typePeripherique) {
    return api.get(`/articles/peripheriques/type/${typePeripherique}`)
  },

  // Telephones
  getAllTelephones() {
    return api.get('/articles/telephones/all')
  },

  getTelephonesByMarque(marque) {
    return api.get(`/articles/telephones/marque/${marque}`)
  },

  getArticlesByCustomCategory(categoryName) {
    return api.get(`/articles/custom/${categoryName}`)
  }
}

export default ArticleService

File: app-back/src/main/java/com/esb/appback/service/ArticleService.java
package com.esb.appback.service;

import com.esb.appback.model.*;
import com.esb.appback.repository.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import java.util.List;
import java.util.Optional;
import java.util.Map;
import java.util.stream.Collectors;
import com.esb.appback.dto.ArticleStatsDTO;

@Service
public class ArticleService {

    @Autowired
    private ArticleRepository articleRepository;

    @Autowired
    private ConsommableRepository consommableRepository;

    @Autowired
    private CartoucheRepository cartoucheRepository;

    @Autowired
    private MeubleRepository meubleRepository;

    @Autowired
    private BureauRepository bureauRepository;

    @Autowired
    private ChaiseRepository chaiseRepository;

    @Autowired
    private EtagereIndustrielleRepository etagereRepository;

    @Autowired
    private ModeleImprimanteRepository imprimanteRepository;

    // Generic Article operations
    public List<Article> getAllArticles() {
        return articleRepository.findAll();
    }

    public Optional<Article> getArticleById(Long id) {
        return articleRepository.findById(id);
    }

    public Optional<Article> getArticleByCodeBarre(String codeBarre) {
        return articleRepository.findByCodeBarre(codeBarre);
    }

    public List<Article> getArticlesByType(String type) {
        return articleRepository.findByType(type);
    }

    public List<Article> getArticlesByFamille(String famille) {
        return articleRepository.findByFamille(famille);
    }

    public List<Article> getArticlesByFournisseur(String fournisseur) {
        return articleRepository.findByFournisseur(fournisseur);
    }

    public List<String> getAllTypes() {
        return articleRepository.findAllTypes();
    }

    public List<String> getAllFamilles() {
        return articleRepository.findAllFamilles();
    }

    public Article saveArticle(Article article) {
        // Auto-generate QR code if not already set
        if (article.getQrCode() == null || article.getQrCode().isEmpty()) {
            String code = article.getCodeBarre();
            if (code != null && !code.isEmpty()) {
                article.setQrCode(generateQRCode(code));
            } else {
                article.setQrCode(""); // or skip
            }
        }
        return articleRepository.save(article);
    }

    private String generateQRCode(String data) {
        // Using goqr.me API to generate QR codes
        // Format: https://api.qrserver.com/v1/create-qr-code/?size=200x200&data=<data>
        return "https://api.qrserver.com/v1/create-qr-code/?size=200x200&data=" + encodeData(data);
    }

    private String encodeData(String data) {
        try {
            return java.net.URLEncoder.encode(data, "UTF-8");
        } catch (java.io.UnsupportedEncodingException e) {
            return data;
        }
    }

    public void deleteArticle(Long id) {
        articleRepository.deleteById(id);
    }

    // Category-specific operations
    public List<Consommable> getAllConsommables() {
        return consommableRepository.findAll();
    }

    public List<Cartouche> getAllCartouches() {
        return cartoucheRepository.findAll();
    }

    public List<Cartouche> getCartouchesByColor(String couleur) {
        return cartoucheRepository.findByCouleur(couleur);
    }

    public List<Meuble> getAllMeubles() {
        return meubleRepository.findAll();
    }

    public List<Meuble> getMeublesByMaterial(String materiau) {
        return meubleRepository.findByMateriau(materiau);
    }

    public List<Bureau> getAllBureaux() {
        return bureauRepository.findAll();
    }

    public List<Chaise> getAllChaises() {
        return chaiseRepository.findAll();
    }

    public List<EtagereIndustrielle> getAllEtageres() {
        return etagereRepository.findAll();
    }

    public List<ModeleImprimante> getAllImprimantes() {
        return imprimanteRepository.findAll();
    }

    public List<ModeleImprimante> getImprimantesByMarque(String marque) {
        return imprimanteRepository.findByMarque(marque);
    }

    public ArticleStatsDTO getArticleStats() {
        List<Article> articles = articleRepository.findAll();

        long totalArticles = articles.size();
        double totalValue = articles.stream()
                .filter(a -> a.getPrix() != null && a.getQuantite() != null)
                .mapToDouble(a -> a.getPrix() * a.getQuantite())
                .sum();

        Map<String, Long> countByType = articles.stream()
                .filter(a -> a.getType() != null)
                .collect(Collectors.groupingBy(Article::getType, Collectors.counting()));

        Map<String, Long> countByFamille = articles.stream()
                .filter(a -> a.getFamille() != null)
                .collect(Collectors.groupingBy(Article::getFamille, Collectors.counting()));

        long lowStockCount = articles.stream()
                .filter(a -> a.getQuantite() != null && a.getQuantite() < 5)
                .count();

        return ArticleStatsDTO.builder()
                .totalArticles(totalArticles)
                .totalValue(totalValue)
                .countByType(countByType)
                .countByFamille(countByFamille)
                .lowStockCount(lowStockCount)
                .build();
    }
}
